### PR TITLE
Merge 1.4 master

### PR DESCRIPTION
**Pull request by solivier**
http://core-ci.akeneo.com/job/dci_behat/232 CE-ORM
http://core-ci.akeneo.com/job/dci_behat/236 CE-ODM

### CODE REVIEW

**Comment by juliensnz** [comment]
GTM

File: features/Context/Page/Base/Grid.php [hunk 1]
```diff
@@ -51,32 +52,32 @@ public function __construct($session, $pageFactory, $parameters = [])
...
+                        return $grid;
+                    }
                 }
-            }
 
-            return false;
-        }, 'No visible grids found');
+                return false;
+            },
+            20,
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: please remove spin timeout on master

File: features/Context/Page/Base/Grid.php [hunk 2]
```diff
@@ -488,48 +493,23 @@ public function countColumns()
...
-        }
-
-        return $sorter;
+        $header = $this->getColumnHeader($columnName);
+
+        return $this->spin(
+            function () use ($header) {
+                return $header->find('css', 'a');
+            },
+            20,
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: please remove

File: features/Context/Page/Base/Grid.php [hunk 3]
```diff
@@ -614,48 +594,30 @@ public function hideFilter($filterName)
...
-                if ($resetBtn) {
-                    $resetBtn->click();
-
-                    return true;
-                }
+        $resetBtn = $this->spin(function () {
+            return $this
+                ->getElement('Grid toolbar')
+                ->find('css', sprintf('a:contains("%s")', 'Reset'));
+        }, 20, 'Reset button not found');
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 4]
```diff
@@ -614,48 +594,30 @@ public function hideFilter($filterName)
...
-            $this->spin(function () {
-                $refreshBtn = $this
-                    ->getElement('Grid toolbar')
-                    ->find('css', sprintf('a:contains("%s")', 'Refresh'));
-                $refreshBtn->click();
+        $refreshBtn = $this->spin(function () {
+            return $this
+                ->getElement('Grid toolbar')
+                ->find('css', sprintf('a:contains("%s")', 'Refresh'));
+        }, 20, 'Refresh button not found');
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 5]
```diff
@@ -758,28 +718,16 @@ protected function deactivateFilter($filterName)
...
-                    ->getElement('Manage filters')
-                    ->find('css', sprintf('label:contains("%s")', $filterName));
-
-                if ($filterElement) {
-                    $filterElement->click();
+        $filterElement = $this->spin(function () use ($filterName) {
+            return $this
+                ->getElement('Manage filters')
+                ->find('css', sprintf('label:contains("%s")', $filterName));
+        }, 20, sprintf('Impossible to activate filter "%s"', $filterName));
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 6]
```diff
@@ -787,18 +735,13 @@ protected function clickOnFilterToManage($filterName)
...
-            $this->spin(function () {
-                $filterList = $this
-                    ->getElement('Filters')
-                    ->find('css', 'a#add-filter-button');
-                $filterList->click();
+        $filterList = $this->spin(function () {
+            return $this
+                ->getElement('Filters')
+                ->find('css', '#add-filter-button');
+        }, 20, 'Impossible to find filter list');
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 7]
```diff
@@ -807,43 +750,19 @@ protected function clickFiltersList()
...
-                    throw new \InvalidArgumentException(
-                        sprintf('Couldn\'t find row for "%s"', $value)
-                    );
-                }
-
-                $checkbox = $row->find('css', 'input[type="checkbox"]');
+        $row      = $this->getRow($value);
+        $checkbox = $this->spin(function () use ($row) {
+            return $row->find('css', 'input[type="checkbox"]');
+        }, 20, sprintf('Couldn\'t find a checkbox for row "%s"', $value));
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 8]
```diff
@@ -1112,14 +1024,14 @@ public function sequentialEdit()
...
     public function selectAll()
     {
-        if (!$allBtn = $this->getDropdownSelector()->find('css', 'button:contains("All")')) {
-            throw new \InvalidArgumentException('"All" button on dropdown row selector not found');
-        }
+        $selector = $this->getDropdownSelector();
+
+        $allBtn = $this->spin(function () use ($selector) {
+            return $selector->find('css', 'button:contains("All")');
+        }, 20, '"All" button on dropdown row selector not found');
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 9]
```diff
@@ -1143,36 +1055,33 @@ public function selectNone()
...
     protected function getDropdownSelector()
     {
-        if (!$dropdown = $this->getElement('Grid')->find('css', 'th div.btn-group')) {
-            throw new \InvalidArgumentException('Grid dropdown row selector not found');
-        }
-
-        return $dropdown;
+        return $this->spin(function () {
+            return $this->getElement('Grid')->find('css', 'th .btn-group');
+        }, 20, 'Grid dropdown row selector not found');
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 10]
```diff
@@ -1143,36 +1055,33 @@ public function selectNone()
...
     protected function clickOnDropdownSelector($item)
     {
-        if (!$dropdown = $this->getDropdownSelector()->find('css', 'button.dropdown-toggle')) {
-            throw new \InvalidArgumentException('Dropdown row selector not found');
-        }
+        $selector = $this->getDropdownSelector();
+
+        $dropdown = $this->spin(function () use ($selector) {
+            return $selector->find('css', 'button.dropdown-toggle');
+        }, 20, 'Dropdown row selector not found');
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

File: features/Context/Page/Base/Grid.php [hunk 11]
```diff
@@ -1143,36 +1055,33 @@ public function selectNone()
...
 
         $dropdown->click();
-        if (!$listItem = $dropdown->getParent()->find('css', sprintf('li:contains("%s") a', $item))) {
-            throw new \InvalidArgumentException(sprintf('Item "%s" of dropdown row selector not found', $item));
-        }
+
+        $listItem = $this->spin(function () use ($dropdown, $item) {
+            return $dropdown->getParent()->find('css', sprintf('li:contains("%s") a', $item));
+        }, 20, sprintf('Item "%s" of dropdown row selector not found', $item));
```
Comment (grena) [inline features/Context/Page/Base/Grid.php]: remove timeout

**Comment by grena** [comment]
Second argument to remove in spin method (timeout duration no longer exists on master), then GTM!

**Pull request merged**